prime-router/docs/proposals/0020-universal-metadata-search/0020-universal-metadata-search.md

```diff
@@ -0,0 +1,447 @@
+leveraged to accomplish the same goal.  This solution supports normalizing any FHIR resource type and is a more extensible
+solution than something that is custom.
+
+This is an open source server that accepts FHIR data and then provides various mechanisms for searching that data and
+could serve as the tool used for querying against the items that flow through ReportStream.
+
+![search](./search-patient.png)
+![results](./search-results.png)
+![db](./resource-in-db.png)
+
+This approach would entail spinning up a HAPI FHIR server and then accessing the data via the defined API endpoints
+
+Pros:
+
+- Is relatively easy to get started
+- Out of the box search functionality is robust
+- Provides fantastic support for arbitrary searches
+
+Cons:
+
+- Out of the box, `type: "message"` was not accepted by the server; we would likely need to fork or at least figure out to
+get it accepted
+- Exposing functionality not necessarily part of HAPI FHIR server is more difficult
+- At least locally, searches were pretty slow
+- Increases the deployment complexity as another system and infrastructure needs to be maintained
+- Linking back to the lineage data will be complicated/inefficient
+
+
+#### Azure cognitive search
+
+Microsoft has its own custom search indexing offering that could be a potential solution.
+
+Pros:
+
+- Supposedly has neat features "powered" by AI
+- Exposes a REST interface
+
+Cons:
+
+- Cannot be run locally and would require a dev instance
+- Locks ReportStream into azure
+- Linking back to the lineage data will be complicated/inefficient (see the second con for Elasticsearch)
+
+## Recommendation
+
+Overall, the exact requirements around search are still fuzzy; there are a few things that receivers and senders are
+definitely interested in knowing around how data is getting routed (this is also what someone debugging the system is
+interested in), but items such as what is valuable to search in the metadata is still an unknown.
+
+Due to the ambiguity, the recommended approach is:
+
+- track the origin report/items to the terminal report/items in a new column and table
+  - Specifically, we would insert a new row whenever there is no next action to be performed
+- Create a new metadata table with a JSONB column for storing the metadata
+  - After creating the internal FHIR bundle that the UP will operate on, insert that into the metadata table with 
+  origin report and index
+
+This is the simplest approach and requires no new infrastructure or concepts while still maintaining flexibility moving
+forward as requirements evolve.  The largest cons of this solution are mostly ones of scale and monitoring could be
+configured such that the team would have an early warning that one of the other more involved solutions spelled out 
+here need to be adopted
+
+## Searches to be split into other tickets
+
+- Returning a list of "undelivered" reports
```

Comment:
I worry about us not tackling the delivered/undelivered aspect in this ticket. It seems like we could put a lot of work going down one route just to do this research and decide another makes more sense.

Reply:
I'll update this in the doc too, but delivered reports is definitely in scope and covered by this proposal.  I think agree and will include capturing "undelivered" reports too.  The original reason to separate that out was that this data does get captured in the `action` table (action is `send_warning`), but the report id is in a raw string in the action param and results and not easily searchable.  We would need to figure out how to adjust the schema to surface that info in a more queryable fashion